### PR TITLE
LF: ad-hoc data struct to store LF Struct like data

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
A `Struct[+X]` is a list of pair `(FieldName, X)` ordered by first component.
    
We use this data-structure to represent `TStruct` LF type.
    
It will be used in upcomming PRs to sorted field in Struct values.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala [hunk 1]
```diff
@@ -305,50 +305,49 @@ class TypingSpec extends WordSpec with TableDrivenPropertyChecks with Matchers {
...
-        E"Λ (σ : ⋆ → ⋆). λ (e : Mod:R σ) → (( Mod:R @σ (e).f2 ))",
-        E"Λ (σ : ⋆). λ (e : Mod:R σ) → (( Mod:R @σ (e).f3 ))",
-        // ExpRecUpdate
-        E"Λ (σ : ⋆). λ (e : Mod:R σ) (e₂ : List σ) → (( Mod:R @σ { e  with f3 = e₂ } ))",
-        E"Λ (σ : ⋆). λ (e : Mod:R σ) (e₂ : Bool) → (( Mod:R @σ { e  with f2 = e₂ } ))",
-        // ExpVarCon
-        E"Λ (σ : ⋆ → ⋆). λ (e : σ) → (( Mod:Tree:Leaf @σ e ))",
-        E"Λ (σ : ⋆). λ (e : σ) → (( Mod:Tree @σ Cons @σ [e] (Nil @σ) ))",
-        // ExpStructCon
+//        E"x",
```
Comment (cocreature) [inline daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala:308]: What’s up with all of this?

Comment (remyhaemmerle-da) [reply daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala]: fixed.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Struct.scala [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+package com.daml.lf
+package data
+
+import com.daml.lf.data.Ref.Name
+
+final case class Struct[+X] private (sortedFields: ImmArray[(Ref.Name, X)]) extends NoCopy {
+
+  def lookup(name: Ref.Name): Option[X] =
+    sortedFields.find(_._1 == name).map(_._2)
```
Comment (cocreature) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Struct.scala:12]: reminder to myself and others we used to do the linear lookup before as well.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala [hunk 3]
```diff
@@ -26,7 +26,10 @@ private[validation] object TypeTraversable {
         f(body)
         ()
       case TStruct(fields) =>
-        fields.values.foreach(f)
+        fields.values.foreach { x =>
```
Comment (cocreature) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala:29]: why is this better (or maybe even necessary now)?

File: daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/TypeOrdering.scala [hunk 4]
```diff
@@ -49,6 +50,12 @@ object TypeOrdering extends Ordering[Ast.Type] {
         throw new IllegalArgumentException(s"cannot compare types $typ")
     }
 
+  private[this] def compareNames(xs: Iterator[FieldName], ys: Iterator[FieldName]) = {
```
Comment (cocreature) [inline daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/TypeOrdering.scala:53]: I assume this is faster than what we did before? Doesn’t seem simpler.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Struct.scala [hunk 5]
```diff
@@ -0,0 +1,48 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf
+package data
+
+import com.daml.lf.data.Ref.Name
+
+final case class Struct[+X] private (sortedFields: ImmArray[(Ref.Name, X)]) extends NoCopy {
```
Comment (cocreature) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Struct.scala:9]: This currently doesn’t appear to enforce uniqueness of fields. Given that we rely on this and have a lookup method, should it?

**Review by cocreature — APPROVED** [review]
Thanks!

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Struct.scala (hunk 5 above)
Comment (cocreature) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Struct.scala:9]: I’ll leave it up to you, I don’t feel particularly strongly here.

**Pull request merged**